testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java

```diff
@@ -0,0 +1,112 @@
+    private final List<ProcessedRecord> processedRecords =
+            Collections.synchronizedList(new ArrayList<>());
+
+    @Override
+    public void onProduce(ProducedRecord record) {
+        producedRecords.add(record);
+    }
+
+    @Override
+    public void onProcess(ProcessedRecord record) {
+        processedRecords.add(record);
+    }
+
+    @Override
+    public void doAssert() {
+        Map<String, List<TestTask>> perKeyProducedRecords = new HashMap<>();
+        Map<String, List<TestTask>> perKeyProcessedRecords = new HashMap<>();
+        Map<TestTask, Long> taskToOffset = new HashMap<>();
+
+        for (ProducedRecord record : producedRecords) {
+            taskToOffset.put(record.task(), record.offset());
+            perKeyProducedRecords.computeIfAbsent(record.key(),
+                                                  key -> new ArrayList<>()).add(record.task());
+        }
+
+        for (ProcessedRecord record : processedRecords) {
+            perKeyProcessedRecords.computeIfAbsent(record.key(),
+                                                   key -> new ArrayList<>()).add(record.task());
+        }
+
+        for (Entry<String, List<TestTask>> entry : perKeyProducedRecords.entrySet()) {
+            String key = entry.getKey();
+            List<TestTask> produced = entry.getValue();
+            List<TestTask> processed = perKeyProcessedRecords.get(key);
+
+            assertNotNull(processed);
+            assertOrdering(taskToOffset, produced, processed);
+        }
+    }
+
+    static void assertOrdering(Map<TestTask, Long> taskToOffset,
+                               List<TestTask> produced,
+                               List<TestTask> processed) {
+        Deque<TestTask> excludeReprocess = new ArrayDeque<>();
+
+        TestTask headTask = processed.get(0);
+        excludeReprocess.addLast(headTask);
+        long currentOffset = taskToOffset.get(headTask);
+
+        long committed = -1L;
+        for (int i = 1; i < processed.size(); i++) {
+            TestTask task = processed.get(i);
+            long offset = taskToOffset.get(task);
+
+            if (offset < committed) {
+                fail("offset cannot be regressed beyond committed offset");
+            }
+            if (offset <= currentOffset) {
+                // offset regression implies reprocessing from committed offset happened
+                committed = offset;
+
+                // rewind records to committed offset
+                TestTask last;
+                while ((last = excludeReprocess.peekLast()) != null) {
+                    if (taskToOffset.get(last) == offset) {
+                        break;
+                    }
+                    excludeReprocess.removeLast();
+                }
+            } else {
+                excludeReprocess.add(task);
+            }
+            currentOffset = offset;
+        }
+
+        assertTrue(produced.equals(new ArrayList<>(excludeReprocess)));
```

Comment:
why not just `assertEquals`?

Reply:
this reason
https://github.com/line/decaton/pull/18#discussion_r394786361

Reply:
I see. then
- as a hint for assertion we would add on top `assertEquals` for their size?
- `//noinspection SimplifiableJUnitAssertion` please